### PR TITLE
Fix `SVGCircuit()` for `cirq.Circuit` with empty moments

### PR DESCRIPTION
**Pull request by freyam**
This PR aims to solve a bug with `SVGCircuit()`.

- [x] **[BUG]** Take care of the SVG circuits when the `cirq.Moment()` is empty (Currently, it fails).
    - We solved this by replicating the standard Text Representation for circuits consisting of empty moments.
    - Added a suitable test to verify results.
    - In the case of empty moments, we show a blank line.

This is part of the https://github.com/quantumlib/Cirq/issues/4499 which plans on solving some bugs in the `SVGCircuit` class and finally paving the way to integrate the SVGCircuit into `cirq.AbstractCircuit` as `_repr_svg_`. 

I have added several TODOs in the `svg.py` which will be addressed once this PR is merged.

## Demo

Relevant Notebook: https://gist.github.com/vanhalen42/80cf3551bef7fcfb6e9992b88ae5f9e5

```py
a, b, c = cirq.LineQubit.range(3)
circuit2=cirq.Circuit()
circuit2.append(cirq.Moment(cirq.CNOT(a, b)))
circuit2.append(cirq.Moment(cirq.CZ(b, c)))
circuit2.append(cirq.Moment()) # Empty moment included in the circuit
circuit2.append(cirq.Moment(cirq.SWAP(a, c)))
circuit2.append(cirq.Moment(cirq.PhasedXPowGate(exponent=0.123, phase_exponent=0.456).on(c)))
circuit2.append(cirq.Z(a))
circuit2.append(cirq.Moment(cirq.measure(a, b, c, key='z')))
```

### Before

image

### After

image

--------
It's also noticeable when you have an empty moment between 2 non-empty ones.
image

cc @vanhalen42 @tanujkhattar

### CODE REVIEW

**Comment by freyam** [comment]
@googlebot I fixed it.

**Comment by vanhalen42** [comment]
@googlebot I consent.

**Comment by freyam** [comment]
This PR is now ready for review.

File: cirq-core/cirq/circuits/_block_diagram_drawer.py [hunk 1]
```diff
@@ -158,9 +158,9 @@ class BlockDiagramDrawer:
     """Aligns text and curve data placed onto an abstract 2d grid of blocks."""
 
     def __init__(self):
-        self._blocks = collections.defaultdict(Block)  # type: Dict[Tuple[int, int], Block]
-        self._min_widths = collections.defaultdict(lambda: 0)  # type: Dict[int, int]
-        self._min_heights = collections.defaultdict(lambda: 0)  # type: Dict[int, int]
+        self._blocks = collections.defaultdict(Block)  # type: dict[tuple[int, int], Block]
```
Comment (mpharrigan) [inline cirq-core/cirq/circuits/_block_diagram_drawer.py:161]: why the change here? I think this is only valid for newer pythons. We still support 3.7 (and maybe 3.6??)

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -258,6 +258,10 @@ def _repr_html_(self) -> str:
             + '</pre>'
         )
 
+    def _repr_svg_(self) -> str:
```
Comment (mpharrigan) [inline cirq-core/cirq/circuits/circuit.py:261]: remove this stub until its implemented

File: cirq-core/cirq/contrib/svg/svg.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 def fixup_text(text: str):
     if '\n' in text:
+        # TODO: Custom MatrixGate
```
Comment (mpharrigan) [inline cirq-core/cirq/contrib/svg/svg.py:14]: nit: link to issues of outstanding svg issues so we can track this todo

File: cirq-core/cirq/contrib/svg/svg.py [hunk 4]
```diff
@@ -25,8 +26,11 @@ def fixup_text(text: str):
 def _get_text_width(t: str) -> float:
     t = fixup_text(t)
     tp = matplotlib.textpath.TextPath((0, 0), t, size=14, prop=FONT)
-    bb = tp.get_extents()
-    return bb.width + 10
+    try:
+        bb = tp.get_extents()
+        return bb.width + 10
+    except:
+        return 21
```
Comment (mpharrigan) [inline cirq-core/cirq/contrib/svg/svg.py:33]: where does this magic number come from? consider adding a comment or factoring out into a meaningfully-named global constant

File: cirq-core/cirq/contrib/svg/svg.py [hunk 5]
```diff
@@ -270,11 +270,15 @@ def _repr_svg_(self) -> str:
         # coverage: ignore
         _validate_circuit(self.circuit)
         tdd = self.circuit.to_text_diagram_drawer(transpose=False)
+        if len(tdd.horizontal_lines) == 0:
+            return '<svg></svg>'
```
Comment (mpharrigan) [inline cirq-core/cirq/contrib/svg/svg.py:274]: add a comment describing this case

File: cirq-core/cirq/contrib/svg/svg.py [hunk 6]
```diff
@@ -270,11 +270,15 @@ def _repr_svg_(self) -> str:
...
         return tdd_to_svg(tdd)
 
 
 def circuit_to_svg(circuit: 'cirq.Circuit') -> str:
     """Render a circuit as SVG."""
     _validate_circuit(circuit)
     tdd = circuit.to_text_diagram_drawer(transpose=False)
+    if len(tdd.horizontal_lines) == 0:
```
Comment (mpharrigan) [inline cirq-core/cirq/contrib/svg/svg.py:282]: why do we need this again?

File: cirq-core/cirq/contrib/svg/svg_test.py [hunk 7]
```diff
@@ -36,6 +36,8 @@ def test_validation():
     with pytest.raises(ValueError):
         circuit_to_svg(cirq.Circuit())
 
-    q0 = cirq.LineQubit(0)
-    with pytest.raises(ValueError):
-        circuit_to_svg(cirq.Circuit([cirq.Moment([cirq.X(q0)]), cirq.Moment([])]))
+
+def test_empty_moments():
```
Comment (mpharrigan) [inline cirq-core/cirq/contrib/svg/svg_test.py:40]: what about a test for a circuit that has non-empty moments and one empty moment? this is the common case after e.g. you run an optimizer to combine gates and it leaves behind empty moments

**Review by mpharrigan — COMMENTED** [review]
Can you add a screenshot of what this looks like? How easy is it to tell an empty moment between two moments from just the two moments with nothing inbetween?

File: cirq-core/cirq/circuits/_block_diagram_drawer.py (hunk 1 above)
Comment (freyam) [reply cirq-core/cirq/circuits/_block_diagram_drawer.py:161]: Yes, you are right.

I had just changed `Dict` to `dict` keeping in mind the current Python guidelines (Python 3.9 calls it `dict` and `tuple`).
I have reverted them for the time being.

File: cirq-core/cirq/contrib/svg/svg.py (hunk 3 above)
Comment (freyam) [reply cirq-core/cirq/contrib/svg/svg.py:14]: Great! Added :heavy_check_mark:

File: cirq-core/cirq/contrib/svg/svg_test.py (hunk 7 above)
Comment (vanhalen42) [reply cirq-core/cirq/contrib/svg/svg_test.py:40]: Oh, that sounds like a good plan. Added.

File: cirq-core/cirq/contrib/svg/svg.py (hunk 6 above)
Comment (vanhalen42) [reply cirq-core/cirq/contrib/svg/svg.py]: In circuits with no non-empty moments, this will return a blank SVG. (this code snippet was redundant as the same code was present in _repr_svg_ ). I have added a comment describing the case.

File: cirq-core/cirq/contrib/svg/svg.py (hunk 4 above)
Comment (vanhalen42) [reply cirq-core/cirq/contrib/svg/svg.py]: The value returned for the CNOT gate in `_get_text_width` is about 21, and this seems like a good enough margin to tell the empty moments apart in the SVG representation.

**Comment by freyam** [comment]
@mpharrigan I have updated the PR's top comment to reflect the changes.

**Comment by freyam** [comment]
@tanujkhattar I believe this PR is ready to be merged. The consequent work can be taken up in a fresh PR.
